src/vec.rs

```diff
@@ -13,32 +13,65 @@ impl<T: PartialOrd> StalinFind<T> for Vec<T> {
     self.is_empty()
   }
 
-  fn stalin(&mut self, i: T, l: usize, r: usize) -> Option<usize>
-    where T: cmp::PartialEq + cmp::PartialOrd {
-    let m = (l + r) / 2;
-    if self[m] == i {
-      Some(m)
-    } else {
-      self.remove(m);
-      if m == 0 {
-        if self.is_empty() {
-          None
-        } else {
-          self.stalin(i, 0, self.len() - 1)
+    fn stalin(&mut self, i: T, l: usize, r: usize) -> Option<usize>
+    where T: cmp::PartialEq + cmp::PartialOrd 
+    {
+        let m = (l + r) / 2;
+        if self[m] == i 
+        {
+            Some(m)
+        }
+        else 
+        {
+            if l != r
+            {
+                if self[m] > i
+                {
+                    self.remove(m);
+                    // a special care is only needed for the left edge 
+                    // since "r" always is always bigger than "m" at least by 1
+                    if m == l { self.stalin(i, l, m) } 
+                    else { self.stalin(i, l, m - 1) }
+                } 
+                else 
+                {
+                    self.remove(m);
+                    self.stalin(i, m, r - 1)
+                }
+            }
+            else { None }
         }
-      } else if self[m - 1] > i {
-        self.stalin(i, l, m - 1)
-      } else {
-        self.stalin(i, m - 1, r - 1)
-      }
     }
-  }
 }
 
 #[cfg(test)]
 mod tests {
   use super::*;
 
+  #[test]
+  fn randomised_test() {
+    // generate a random vector of length 0-n
+    let n = 100; // n was supposed to be a paramer of the function but whatever
+    let mut rng = rand::thread_rng();
+    let sought = rng.gen_range(0, n);
+    let amount = rng.gen_range(0, n);
+    let range = Uniform::new(0, n);
+    let mut vector: Vec<u64> = (0..amount).map(|_| rng.sample(&range)).collect();
+
+    // check if the value is within the array
+    let i = vector.iter().position(|x| x == &sought);
+    if i.is_some() {
+        // IM NOT SURE HERE AS WELL THE LANGUAGE IS STUPID EVERYTHING IS STUPID I HATE IT
```

Comment:
is this inspired by my C++ code comments? ^_^

Reply:
gotta patent the ARG styling pattern egh?